python/ray/experimental/workflow/workflow_manager.py

```diff
@@ -89,23 +88,26 @@ def _resolve_step_inputs(step_inputs: Tuple[RRef, List[RRef], List[RRef]]
 
 class WorkflowStepFunction:
     def __init__(self, func: Callable):
-        def _func(context, task_id, step_inputs):
-            # NOTE: must use 'set_current_store_dir' to ensure that we are
-            # accessing the correct global variable.
-            workflow_context.set_workflow_step_context(context)
-            scope = workflow_context.get_scope()
-            scope.append(task_id)
+        def _func(context, task_id, step_inputs, forward_output_to):
+            # Before running the actual function, we
+            # 1. Setup the workflow context, so we have proper access to
+            #    workflow storage.
+            # 2. Decode step inputs to arguments and keyword-arguments.
+            workflow_context.update_workflow_step_context(context, task_id)
             args, kwargs, resolved_object_refs = _resolve_step_inputs(
                 step_inputs)
             # free references to potentially save memory
             del resolved_object_refs
 
-            output = func(*args, **kwargs)
-            if isinstance(output, Workflow):
-                output = output.execute()
-            return output
+            # Running the actual step function
+            ret = func(*args, **kwargs)
+            # After running the actual step function, we checkpoint the
+            # output. If "forward_output_to" is not None, we forward the
+            # output to the target step, so when resuming the workflow
+            # we can access the output faster.
+            return postprocess_workflow_step(ret, forward_output_to)
 
-        self.func = func
+        self._func = func
```

Comment:
Is it necessary to store func to `self._func`?

Reply:
Yes, we need to store the original function body. This is because during recovery, we require the original function to compute the result.

Reply:
the "transformed" function in WorkflowStepFunction `_func` (not `self._func` sorry the name is confusing) is not suitable for recovery.